### PR TITLE
Add OData writer validation tests for unhappy paths

### PR DESCRIPTION
**Pull request by gathogojr**
### Description

This pulls adds OData writer validation tests for unhappy paths that were mostly untested. Step 1 of 2 towards improving OData writer validation error messages relevant to this reported issue https://github.com/OData/odata.net/issues/2960

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### CODE REVIEW

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs [hunk 1]
```diff
@@ -0,0 +1,3356 @@
...
+            await using (var messageWriter = new ODataMessageWriter(responseMessage, this.messageWriterSettings, this.model))
+            {
+                var exception = await Assert.ThrowsAsync<ODataException>(() => messageWriter.WritePropertyAsync(new ODataProperty
+                {
+                    Name = "Name",
+                    Value = null
+                }));
+
+                Assert.Equal(
+                    Strings.ODataMessageWriter_CannotWriteTopLevelNull,
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs:491]: Why cannot write the Top level 'null'?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs [hunk 2]
```diff
@@ -0,0 +1,3356 @@
...
+                    exception.Message);
+            }
+        }
+
+        [Theory]
+        [InlineData("")]
+        [InlineData(null)]
+        public async Task WriterValidationUtils_PropertiesMustHaveNonEmptyNameAsync(string propertyName)
+        {
+            IODataRequestMessage requestMessage = new InMemoryMessage { Stream = this.stream };
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs:625]: It seems this line is for all test case, maybe it's better to make it as instance field and You don't need to repeat it in every unit test.

**Review by xuzhg — APPROVED** [review]

**Review by habbes — DISMISSED** [review]

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs [hunk 3]
```diff
@@ -0,0 +1,3356 @@
...
+                new TempEntityMetadataBuilder(
+                    resourceId: this.productResource.Id,
+                    mediaResource: productPhotoMediaResource,
+                    eTag: null),
+                "Photo");
+
+            this.messageWriterSettings = new ODataMessageWriterSettings
+            {
+                EnableMessageStreamDisposal = false,
+                Version = ODataVersion.V4,
```
Comment (WanjohiSammy) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs:238]: Will all the tests work with `ODataVersion.V401`

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs (hunk 1 above)
Comment (gathogojr) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs:489]: @xuzhg Based on our logic (and the spec), if a top-level single-valued property is requested but it has a null value, the service should respond with 204 No Content
https://github.com/OData/odata.net/blob/713ea85bf61846a6978acc2a3d1c792e0a9d0af8/src/Microsoft.OData.Core/WriterValidator.cs#L253

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs (hunk 2 above)
Comment (gathogojr) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs]: @xuzhg Refactored such that all tests use helper methods that contain the shared logic

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs [hunk 4]
```diff
@@ -0,0 +1,3356 @@
...
+        }
+
+        [Fact]
+        public async Task WriterValidationUtils_ExpandedLinkWithResourcePayloadAndResourceSetMetadataAsync()
+        {
+            IODataResponseMessage responseMessage = new InMemoryMessage { Stream = this.stream };
+
+            await using (var messageWriter = new ODataMessageWriter(responseMessage, this.messageWriterSettings, this.model))
+            {
+                var exception = await Assert.ThrowsAsync<ODataException>(async () =>
```
[the remark this answers is not in the document]
Comment (gathogojr) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs:875]: @xuzhg Refactored such that all tests use helper methods that contain the shared logic

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs (hunk 3 above)
Comment (gathogojr) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterValidationTests.cs]: @WanjohiSammy Parameterized all tests such that they're ran for both V4 and V401.

**Review by WanjohiSammy — APPROVED** [review]

**Pull request merged**